ni845x_spi.c

```diff
@@ -0,0 +1,501 @@
+                     "or use 'lsusb -v -d 3923:7166 | grep Serial' for USB-8451\n"
+                     "or 'lsusb -v -d 3923:7514 | grep Serial' for USB-8452\n");
+            free(serial_number);
+        } else {
+            msg_pinfo("Could not find any NI USB-845x device!\n");
+        }
+        return 1;
+    } else {
+        // open the SPI config handle
+        tmp = ni845xSpiConfigurationOpen(&configuration_handle);
+        if (tmp != 0) {
+            ni845xStatusToString(tmp, ARRAY_SIZE(error_string_buffer), error_string_buffer);
+            msg_perr("ni845xSpiConfigurationOpen() failed with: %s (%ld)\n",
+                     error_string_buffer,
+                     tmp);
+        }
+    }
+
+    if (serial_number)
+        free(serial_number);
+
+    if (usb8452_spi_set_io_voltage(IO_voltage_mV) < 0) {
+        return 1; // no alert here usb8452_spi_set_io_voltage already printed that
+    }
+
+    speed_str = extract_programmer_param("spispeed");
+    if (speed_str) {
+        spi_speed_KHz = atoi(speed_str);
+        free(speed_str);
+    }
+
+    if (ni845x_spi_set_speed(spi_speed_KHz)) {
+        msg_perr("Unable to set SPI speed\n");
+        return 1;
+    }
+
+    if (register_shutdown(ni845x_spi_shutdown, NULL))
+        return 1;
+    register_spi_master(&spi_programmer_ni845x);
+
+    return 0;
+}
+
+static int ni845x_spi_shutdown(void *data)
+{
+    int32 ret = 0;
+    if (configuration_handle != 0) {
+        ret = ni845xSpiConfigurationClose(configuration_handle);
+        if (ret) {
+            ni845xStatusToString(ret, ARRAY_SIZE(error_string_buffer), error_string_buffer);
+            msg_perr("ni845xSpiConfigurationClose failed with: %s (%ld)\n", error_string_buffer, ret);
+        }
+    }
+
+    if (device_handle != 0) {
+        ret = ni845xClose(device_handle);
+        if (ret) {
+            ni845xStatusToString(ret, ARRAY_SIZE(error_string_buffer), error_string_buffer);
+            msg_perr("ni845xSpiConfigurationClose failed with: %s (%ld)\n", error_string_buffer, ret);
+        }
+    }
+    return 0;
+}
+
+static int ni845x_spi_transmit(struct flashctx *flash,
+                                    unsigned int write_cnt,
+                                    unsigned int read_cnt,
+                                    const unsigned char *write_arr,
+                                    unsigned char *read_arr)
+{
+    uInt32 read_size = 0;
+    uInt8 *transfer_buffer = NULL;
+    int32 ret = 0;
+
+    transfer_buffer = calloc(write_cnt + read_cnt, sizeof(uInt8));
+    if (transfer_buffer == NULL) {
+        msg_gerr("Memory allocation failed!\n");
+        msg_cinfo("FAILED.\n");
+        return -1;
+    }
+
+    memcpy(transfer_buffer, write_arr, write_cnt);
+
+    ret = ni845xSpiWriteRead(device_handle,
+                                   configuration_handle,
+                                   (uInt32)(write_cnt + read_cnt),
+                                   (uInt8*)transfer_buffer,
```

Comment:
No need for the cast to uInt8 since it was already declared as such.

Reply:
True, fixed.